File: java/4-tidy-first/src/test/java/com/tidyfirst/EventTicketServiceTest.java
package com.tidyfirst;

import org.junit.jupiter.api.Test;

import java.util.HashMap;

import static org.assertj.core.api.Assertions.assertThat;

class EventTicketServiceTest {

    private NotificationService notificationService = new NotificationService();
    private EventTicketService service = new EventTicketService(notificationService);

    private HashMap<String, Object> options(double basePrice, int quantity, boolean isEarlyBird, boolean isVip) {
        HashMap<String, Object> opts = new HashMap<>();
        opts.put("basePrice", basePrice);
        opts.put("quantity", quantity);
        opts.put("isEarlyBird", isEarlyBird);
        opts.put("isVip", isVip);
        return opts;
    }

    @Test
    void calculatesNormalPriceWithTax() {
        assertThat(service.calculateTicketPrice(options(100, 1, false, false))).isEqualTo(119.0);
    }

    @Test
    void returnsZeroForZeroQuantity() {
        assertThat(service.calculateTicketPrice(options(100, 0, false, false))).isEqualTo(0);
    }

    @Test
    void appliesEarlyBirdDiscount() {
        assertThat(service.calculateTicketPrice(options(100, 1, true, false))).isEqualTo(101.15);
    }

    @Test
    void appliesVipSurcharge() {
        assertThat(service.calculateTicketPrice(options(100, 1, false, true))).isEqualTo(178.5);
    }

    @Test
    void appliesGroupDiscountForTenOrMoreTickets() {
        assertThat(service.calculateTicketPrice(options(100, 10, false, false))).isEqualTo(1011.5);
    }

    @Test
    void combinesEarlyBirdAndVip() {
        assertThat(service.calculateTicketPrice(options(100, 1, true, true))).isEqualTo(151.73);
    }

    @Test
    void combinesEarlyBirdAndGroupDiscount() {
        assertThat(service.calculateTicketPrice(options(100, 10, true, false))).isEqualTo(859.78);
    }

    @Test
    void combinesVipAndGroupDiscount() {
        assertThat(service.calculateTicketPrice(options(100, 10, false, true))).isEqualTo(1517.25);
    }

    @Test
    void combinesAllDiscounts() {
        assertThat(service.calculateTicketPrice(options(100, 10, true, true))).isEqualTo(1289.66);
    }

    @Test
    void generatesConfirmationMessage() {
        String msg = service.generateConfirmationMessage("Rock Concert", 2, 238.0);
        assertThat(msg).contains("Rock Concert", "2 ticket(s)", "238.00");
    }

    @Test
    void generatesReceiptMessage() {
        String msg = service.generateReceiptMessage("Rock Concert", 2, 238.0, "Max Mustermann");
        assertThat(msg).contains("Max Mustermann", "2x Rock Concert", "238.00");
    }

    @Test
    void generatesSummaryMessage() {
        String msg = service.generateSummaryMessage("Rock Concert", "2026-04-15", "Hamburg Arena");
        assertThat(msg).contains("Rock Concert", "2026-04-15", "Hamburg Arena");
    }

    @Test
    void sendsNotificationViaNotificationService() {
        service.sendBookingConfirmation("test@example.com", "Rock Concert", 2, 238.0);
        assertThat(notificationService.getSentMessages()).hasSize(1);
        assertThat(notificationService.getSentMessages().get(0)).contains("test@example.com", "Rock Concert");
    }
}

File: typescript/4-tidy-first/src/event-ticket-service.test.ts
import { describe, it, expect } from "vitest";
import { EventTicketService, NotificationService } from "./event-ticket-service";

describe("EventTicketService", () => {
  function createService() {
    const notificationService = new NotificationService();
    const service = new EventTicketService(notificationService);
    return { service, notificationService };
  }

  describe("calculateTicketPrice", () => {
    it("calculates normal price with tax", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 1, isEarlyBird: false, isVip: false });
      expect(price).toBe(119.0);
    });

    it("returns 0 for zero quantity", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 0 });
      expect(price).toBe(0);
    });

    it("applies early bird discount", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 1, isEarlyBird: true, isVip: false });
      expect(price).toBe(101.15);
    });

    it("applies VIP surcharge", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 1, isEarlyBird: false, isVip: true });
      expect(price).toBe(178.5);
    });

    it("applies group discount for 10+ tickets", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 10, isEarlyBird: false, isVip: false });
      expect(price).toBe(1011.5);
    });

    it("combines early bird and VIP", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 1, isEarlyBird: true, isVip: true });
      expect(price).toBe(151.73);
    });

    it("combines early bird and group discount", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 10, isEarlyBird: true, isVip: false });
      expect(price).toBe(859.78);
    });

    it("combines VIP and group discount", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 10, isEarlyBird: false, isVip: true });
      expect(price).toBe(1517.25);
    });

    it("combines all: early bird, VIP, group", () => {
      const { service } = createService();
      const price = service.calculateTicketPrice({ basePrice: 100, quantity: 10, isEarlyBird: true, isVip: true });
      expect(price).toBe(1289.66);
    });
  });

  describe("message generation", () => {
    it("generates confirmation message", () => {
      const { service } = createService();
      const msg = service.generateConfirmationMessage("Rock Concert", 2, 238.0);
      expect(msg).toContain("Rock Concert");
      expect(msg).toContain("2 ticket(s)");
      expect(msg).toContain("238.00");
    });

    it("generates receipt message", () => {
      const { service } = createService();
      const msg = service.generateReceiptMessage("Rock Concert", 2, 238.0, "Max Mustermann");
      expect(msg).toContain("Max Mustermann");
      expect(msg).toContain("2x Rock Concert");
      expect(msg).toContain("238.00");
    });

    it("generates summary message", () => {
      const { service } = createService();
      const msg = service.generateSummaryMessage("Rock Concert", "2026-04-15", "Hamburg Arena");
      expect(msg).toContain("Rock Concert");
      expect(msg).toContain("2026-04-15");
      expect(msg).toContain("Hamburg Arena");
    });
  });

  describe("sendBookingConfirmation", () => {
    it("sends notification via notification service", () => {
      const { service, notificationService } = createService();
      service.sendBookingConfirmation("test@example.com", "Rock Concert", 2, 238.0);
      expect(notificationService.sentMessages).toHaveLength(1);
      expect(notificationService.sentMessages[0]).toContain("test@example.com");
      expect(notificationService.sentMessages[0]).toContain("Rock Concert");
    });
  });
});
